extensions/amp-app-banner/1.0/component/android.js

```diff
@@ -0,0 +1,106 @@
+import {docInfoService} from '#preact/services/document';
+import {platformService} from '#preact/services/platform';
+import {timerService} from '#preact/services/timer';
+import {urlService} from '#preact/services/url';
+import {viewerService} from '#preact/services/viewer';
+import {xhrService} from '#preact/services/xhr';
+
+import {user} from '#utils/log';
+
+import {openWindowDialog} from '../../../../src/open-window-dialog';
+
+const OPEN_LINK_TIMEOUT = 1500;
+
+/**
+ * @return {{openOrInstall: function(): void, promise: Promise<Response>}|null}
+ */
+export function getAndroidAppInfo() {
+  const win = self.window;
+  // We want to fallback to browser builtin mechanism when possible.
+  const isChromeAndroid =
+    platformService.isAndroid() && platformService.isChrome();
+  const isProxyOrigin = urlService.isProxyOrigin(win.location);
+  const canShowBuiltinBanner =
+    !isProxyOrigin && !viewerService.isEmbedded() && isChromeAndroid;
+
+  if (canShowBuiltinBanner) {
+    user().info(
+      'BENTO-APP-BANNER',
+      'Browser supports builtin banners. Not rendering amp-app-banner.'
+    );
+    return null;
+  }
+
+  const manifestLink = win.document.head.querySelector(
+    'link[rel=manifest],link[rel=origin-manifest]'
+  );
+
+  const missingDataSources = !manifestLink;
+  if (missingDataSources) {
+    return null;
+  }
+
+  const manifestHref = manifestLink.getAttribute('href');
+
+  urlService.assertHttpsUrl(manifestHref, undefined, 'manifest href');
+
+  const promise = xhrService.fetchJson(manifestHref).then(parseManifest);
+  return {
+    promise,
+    openOrInstall: () => {
+      promise.then((manifest) => {
+        if (!manifest) {
+          return;
+        }
+        const {installAppUrl, openInAppUrl} = manifest;
+        timerService.delay(() => {
+          window.top.location.assign(installAppUrl);
+        }, OPEN_LINK_TIMEOUT);
+        openWindowDialog(window, openInAppUrl, '_top');
+      });
+    },
+  };
+}
+
+/**
+ * @param {object} manifestJson
+ * @return {{installAppUrl: string, openInAppUrl: string}|null}
+ */
+function parseManifest(manifestJson) {
+  const apps = manifestJson['related_applications'];
+  if (!apps) {
+    user().warn(
+      'BENTO-APP-BANNER',
+      'related_applications is missing from manifest.json file: %s'
+    );
+    return null;
+  }
+
+  for (let i = 0; i < apps.length; i++) {
+    const app = apps[i];
+    if (app['platform'] === 'play') {
+      const installAppUrl = `https://play.google.com/store/apps/details?id=${app['id']}`;
+      const openInAppUrl = getAndroidIntentForUrl(app['id']);
+      return {installAppUrl, openInAppUrl};
+    }
+  }
```

Comment:
What might be a bit more declarative would be to use the `find` method on Array.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

If you like how you have it you don't need to. But I think it might be good to do something like
```suggestion
  const app = apps.find(({platform}) => platform === 'play');
  
  if(app) {
        const installAppUrl = `https://play.google.com/store/apps/details?id=${app['id']}`;
        const openInAppUrl = getAndroidIntentForUrl(app['id']);
        return {installAppUrl, openInAppUrl};
  }
```

Reply:
Agreed! The code I ported, I left it as-is, but this would be a good improvement; I'll update in both places to improve the overall quality